File: python/euler/bin/p0081.py
# project euler: problem 81

import sys
from collections.abc import Callable
from itertools import accumulate
from typing import TextIO


def compute(fn: Callable[..., int], fh: TextIO) -> str:
    def parse_data(fh: TextIO) -> list[list[int]]:
        return [list(map(int, line.split(','))) for line in fh.read().splitlines()]

    matrix = parse_data(fh)

    # The 'sys.maxsize' is used as sentinel
    prev = [sys.maxsize] + list(accumulate(matrix[0]))
    for work in matrix[1:]:
        work[0:0] = [sys.maxsize]
        for i in range(1, len(work)):
            work[i] += fn(work[i - 1], prev[i])
        prev = work

    return str(prev[-1])


def solve() -> str:
    from euler.lib.resource import asset_file

    fh = asset_file('p081_matrix.txt')
    result = compute(min, fh)
    fh.close()
    return result

File: typescript/euler/bin/p0081.ts
// project euler: problem 81

import { runningReduce } from "std/collections/running_reduce.ts";
import { assetData } from "../lib/asset.ts";
import { range } from "../lib/util.ts";

export const compute = (
  fn: (...valus: number[]) => number,
  data: string,
): string => {
  function parseData(data: string): number[][] {
    function splitLines(str: string): string[] {
      const result = str.split(/\r?\n/);
      if (result.at(-1) === "") {
        return result.slice(0, -1);
      } else {
        return result;
      }
    }

    return splitLines(data).map((x) => x.split(",").map((y) => Number(y)));
  }

  const matrix = parseData(data);

  let prev = [Number.MAX_SAFE_INTEGER].concat(
    runningReduce(matrix[0], (acc, cur) => acc + cur, 0),
  );
  for (const work of matrix.slice(1)) {
    work.unshift(Number.MAX_SAFE_INTEGER);
    for (const i of range(1, work.length)) {
      work[i] += fn(work[i - 1], prev[i]);
    }
    prev = work;
  }

  return String(prev.at(-1));
};

export const solve = (): string => {
  const data = new TextDecoder().decode(assetData("p081_matrix.txt"));
  return compute(Math.min, data);
};
